docs/plugins/testing_workshop_docs/Test-coverage.md

```diff
@@ -0,0 +1,101 @@
+
+## pytest --cov  
+
+To run tests with coverage, run `pytest` and add `--cov` pointing to the module you're covering. There is also an option to generate an html report, which we do in this case. 
+
+    (napari-env) user@directory % 
+    pytest --cov=plugin_tests --cov-report=html .  
+
+Note that the “.” is part of the command.  
+
+    =========================== test session starts ==========================  
+    platform darwin -- Python 3.9.7, pytest-6.2.5, py-1.11.0, pluggy-1.0.0  
+    PyQt 5.15.6 -- Qt runtime 5.15.2 -- Qt compiled 5.15.2  
+    rootdir: qt-4.0.2, napari-plugin-engine-0.2.0, napari-0.4.13, cov-3.0.0  
+    collected 2 items  
+ 
+    src/plugin_tests/_tests/test-reader.py ..                           [100%]  
+ 
+    --------- coverage: platform darwin, python 3.0.7-final-0 --------  
+    Coverage HTML written to dir htmlcov  
+ 
+    =========================== 2 passed in 6.72s ==========================  
+
+This command runs the tests, gets the coverage, and then writes the Coverage HTML report to the htmlcov directory.
+
+There is a large folder (`htmlcov`) in the directory where the tests were run (`plugin_tests`). 
+
+`(napari-env) plugin-tests $ tree -L 1`  
+`.`  
+`├── LICENSE  `  
+`├── MANIFEST.in`  
+`├── README.md`  
+`├── _pycache__`  
+`├── example_func.py`  
+`├── htmlcov		        # <<========================== directory created by pytest-cov`  
+`├── requirements.txt`  
+`├── setup.cfg`  
+`├── setup.py`  
+`├── src`  
+`└── tox.ini`  
+
+
+If we drag the `index.html` file from the list of files in the left panel (to the left of line 32) into a browser, it opens the coverage report as shown below: 
+
+![htmlcov directory](../../images/Test_Coverage_htmlcov_directory.png)
+
+We technically have 100% coverage of `test_widget.py` (the fourth line up from Total) because we didn't write a widget. Since there were zero lines of code in the widget, zero were missed, so we had 100% coverage.  
+
+![Coverage Report](../../images/Coverage_report.png)
+
+We are interested in `_reader.py`. The file containing the reader code has 86% coverage (see below). Clicking ok on the `2 missing` box below highlights the lines that were never run at all. They are highlighted in red (lines 22 and 26): 
+
+![Lines not run highlighted in red](../../images/Lines_not_run_highlighted_in_red.png)
+
+Because we never provided a list of paths and we never ran code that provided a list of paths, we don't know what will happen in that case. We also never ran code that doesn't return a reader. In other words, we never tested the failed cases. We add those tests. The first one is `test_get_reader_pass`. We'll call it with a file that doesn't end in `.npy` and assert that it returned `None`. Then we'll create a second test to call it with a list of paths.
```

Comment:
```suggestion
Because we never provided a list of paths and we never ran code that provided a list of paths, we don't know what will happen in that case. We also never ran code that doesn't return a reader. In other words, we never tested the failed cases. We add can—and should—add those tests. The first one is `test_get_reader_pass`. We'll call it with a file that doesn't end in `.npy` and assert that it returned `None`. Then we'll create a second test to call with a list of paths.
```

Reply:
✓